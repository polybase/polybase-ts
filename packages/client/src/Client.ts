
import { AxiosError, AxiosRequestConfig } from 'axios'
import { createError, createErrorFromAxiosError } from './errors'
import { QueryValue, Request, RequestParams, Sender, SenderResponse, Signer, SignerResponse } from './types'

export interface ClientConfig {
  clientId: string
  baseURL: string
}

type SignatureCache = Record<string, { timeMs: number, sig: SignerResponse }>

export class Client {
  private sender: Sender
  signer?: Signer
  private config?: ClientConfig
  private signatureCache: SignatureCache

  constructor(sender: Sender, signer?: Signer, config?: ClientConfig) {
    this.sender = sender
    this.signer = signer
    this.config = config
    this.signatureCache = {}
  }

  request = (req: Request): ClientRequest => {
    return new ClientRequest(this.sender, {
      url: req.url,
      method: req.method,
      params: parseParams(req.params),
      data: req.data,
    }, this.signatureCache, this.signer, this.config)
  }
}

export class ClientRequest {
  private aborter: AbortController
  private req: Request
  private sender: Sender
  private signer?: Signer
  private config?: ClientConfig
  private signatureCache: SignatureCache

  constructor(sender: Sender, req: Request, signatureCache: SignatureCache, signer?: Signer, config?: ClientConfig) {
    this.aborter = new AbortController()
    this.req = req
    this.sender = sender
    this.signer = signer
    this.config = config
    this.signatureCache = signatureCache
  }

  abort = () => {
    this.aborter.abort()
  }

  /* Sending a request to the server. */
  send = async <T = any>(withAuth: 'none' | 'optional' | 'required', sigExtraTimeMs?: number): Promise<SenderResponse<T>> => {
    try {
      const req = this.req as AxiosRequestConfig
      // Error if we require a signer and its not set
      if (!this.signer && withAuth === 'required') {
        throw createError('request/no-signer', {
          message: 'You must add a .signer() to sign permissioned requests, see: https://polybase.xyz/docs/authentication',
        })
      }

      // Warn if signer is optional, as a signer is recommended
      if (!this.signer && withAuth === 'optional' && process.env.NODE_ENV !== 'production') {
        console.error('Add .signer() to populate ctx.publicKey, see: https://polybase.xyz/docs/authentication')
      }

      if (withAuth !== 'none' && this.signer) {
        const sig = await this.getSignature(sigExtraTimeMs || 0)
        if (sig) {
          if (!req.headers) req.headers = {}
          req.headers['X-Polybase-Signature'] = sig
        }
      }

      const res = await this.sender({
        ...req,
        headers: {
          'X-Polybase-Client': this.config?.clientId ?? 'Polybase',
          ...req.headers,
        },
        baseURL: this.config?.baseURL,
        signal: this.aborter.signal,
      })
      return res
    } catch (e: unknown) {
      if (e && typeof e === 'object' && e instanceof AxiosError) {
        if (e.code === 'ERR_CANCELED') {
          throw createError('request/cancelled')
        }
        throw createErrorFromAxiosError(e)
      }
      throw e
    }
  }

  private getSignature = async (extraTimeMs: number) => {
    if (!this.signer) return ''

    let t: number
    let sig: SignerResponse

    const jsonBody = this.req.data ? JSON.stringify(this.req.data) : ''
    const cachedSignature = this.signatureCache[jsonBody]
    if (cachedSignature && cachedSignature.timeMs > Date.now()) {
      t = cachedSignature.timeMs
      sig = cachedSignature.sig
    } else {
      t = Date.now() + extraTimeMs
      const s = await this.signer(`${t}.${jsonBody}`, this.req)
      if (!s) return null

      sig = s

      if (extraTimeMs > 0) {
        this.signatureCache[jsonBody] = {
          timeMs: t,
          sig: s,
        }
      }
    }

    const h = [
      'v=0',
      `t=${t}`,
      `h=${sig.h}`,
      `sig=${sig.sig}`,
    ]
    if (sig.pk) {
      h.push(`pk=${sig.pk}`)
    }
    return h.join(',')
  }
}

export function parseParams(params?: RequestParams): Record<string, QueryValue | undefined> {
  if (!params) return {}
  return {
    ...params,
    where: params?.where ? JSON.stringify(params?.where) : undefined,
    sort: params?.sort ? JSON.stringify(params?.sort) : undefined,
  }
}
