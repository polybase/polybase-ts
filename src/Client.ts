
import { AxiosRequestConfig } from 'axios'
import { BasicValue, Request, RequestParams, Sender, SenderResponse } from './types'

export class Client {
  private sender: Sender
  private baseURL?: string

  constructor (sender: Sender, baseURL?: string) {
    this.sender = sender
    this.baseURL = baseURL
  }

  request = (req: Request): ClientRequest => {
    return new ClientRequest(this.sender, {
      url: req.url,
      method: req.method,
      params: parseParams(req.params),
      baseURL: this.baseURL,
      data: req.data,
    })
  }
}

export class ClientRequest {
  private aborter: AbortController
  private req: AxiosRequestConfig
  private sender: Sender

  constructor (sender: Sender, req: AxiosRequestConfig) {
    this.aborter = new AbortController()
    this.req = req
    this.sender = sender
  }

  abort = () => {
    this.aborter.abort()
  }

  send = async (): Promise<SenderResponse> => {
    return this.sender({
      ...this.req,
      signal: this.aborter.signal,
    })
  }
}

export function parseParams (params?: RequestParams): Record<string, BasicValue|undefined> {
  if (!params) return {}
  return {
    ...params,
    where: params?.where ? JSON.stringify(params?.where) : undefined,
  }
}
