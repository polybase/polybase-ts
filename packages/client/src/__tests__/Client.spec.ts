import { Client } from '../Client'

test('sign request withAuth: optional', async () => {
  const sender = jest.fn()
  const signer = jest.fn()
  const client = new Client(sender, signer)
  const req = client.request({
    url: '/test',
    method: 'GET',
  })
  await req.send('optional')
  expect(signer).toHaveBeenCalled()
})

test('sign request withAuth: required', async () => {
  const sender = jest.fn()
  const signer = jest.fn()
  const client = new Client(sender, signer)
  const req = client.request({
    url: '/test',
    method: 'GET',
  })
  await req.send('required')
  expect(signer).toHaveBeenCalled()
})

test('errors if no signer withAuth: required', async () => {
  const sender = jest.fn()
  const client = new Client(sender)
  const req = client.request({
    url: '/test',
    method: 'GET',
  })
  await expect(req.send('required')).rejects.toThrowErrorMatchingSnapshot()
})

test('does not sign withAuth: none', async () => {
  const sender = jest.fn()
  const signer = jest.fn()
  const client = new Client(sender, signer)
  const req = client.request({
    url: '/test',
    method: 'GET',
  })
  await req.send('none')
  expect(signer).not.toHaveBeenCalled()
})
