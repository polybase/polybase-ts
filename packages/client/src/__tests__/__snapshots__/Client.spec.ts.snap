// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors if no signer withAuth: required 1`] = `"You must add a .signer() to sign permissioned requests, see: https://polybase.xyz/docs/authentication"`;
